Dependency Updates and ESM Migration

SmushyTaco opened the pull request:
This fixes #68364 to allow for the proper use of control directives.

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR stops the moving of comments and only removes comments that aren't RTLCSS control directives.

## Why?
RTLCSS control directives use comments and if those comments are removed or moved, this will cause this to not function properly. See #68364 for more information.

## How?
The issue is addressed by removing all comments that aren't RTLCSS control directives.

## Testing Instructions
Build this SCSS code:
```scss
.wp-block-create-block-my-block {
  /*!tedgfdfg*/
  /*!rtl:ignore*/
  text-align: left;
  /*!rtl:remove*/
  background: #CCC;
}
```
Only the RTLCSS comments will be preserved as intended.

### Testing Instructions for Keyboard
Not applicable.

## Screenshots or screencast
Not applicable.

SmushyTaco committed:
Fix RTLCSS Support

This solves #68364 to allow for the proper use of control directives.

SmushyTaco committed:
Better Regex

This is done to match the regex found here: https://github.com/MohammadYounes/rtlcss/blob/master/lib/directive-parser.js#L4

gziolo commented:
Thank you for opening this patch. It looks like it will fix the issue. Do you have any thoughts on how the comment could still be removed from the production file, as it's only necessary to instruct `rtlcss`?

SmushyTaco commented:
@gziolo I think I found a better solution. I'll work on writing it and push a new commit once done.

Commit:
Dependency Updates And RTLCSS Fix

Commit:
package.json commits

SmushyTaco commented:
@gziolo I have made the following changes:
1. Updated all dependencies in the `package.json`
2. Replaced `cssnano` with `css-minimizer-webpack-plugin`, this is the recommended way to apply `cssnano` with webpack and fixes the issue with RTLCSS.
3. Deleted the bundled `rtlcss-webpack-plugin` fork and used [this](https://www.npmjs.com/package/@smushytaco/rtlcss-webpack-plugin) one.
4. Replaced the `clean-webpack-plugin` package with [this](https://www.npmjs.com/package/@smushytaco/clean-webpack-plugin) fork.
5. `read-pkg-up` was renamed to `read-package-up` for newer versions so that was changed.
6. `chalk` was replaced with `picocolors` which is 14 times smaller and 2 times as fast.
7. TODO messages were added for changes that need to be made in the future.
8. REGEX expressions were made more concise.

There's only 1 caveat. When building this message is seen:

```
(node:27960) ExperimentalWarning: CommonJS module C:\Users\organ\Documents\GitHub\gutenberg\packages\scripts\utils\package.js is loading ES Module C:\Users\organ\Documents\GitHub\gutenberg\node_modules\read-package-up\index.js using require().
```

I'd like to fix this by converting `@wordpress/scripts` from CommonJS to ESM. Would you be okay with this?

Commit:
Only Use In Production

Commit:
Bump eslint

Commit:
Fix Deprecations And Remove Stale Packages

gziolo commented:
> Replaced cssnano with css-minimizer-webpack-plugin, this is the recommended way to apply cssnano with webpack and fixes the issue with RTLCSS.

I see that the fix boils down to using an existing webpack plugin `css-minimizer-webpack-plugin` that runs `cssnano` after RTL version of the CSS file is created. Cool!

At the moment, CI jobs report multiple issues related to updated npm packages. In practice, it's usually easier to land targeted PRs that solve one problem at a time. For example:

> eslint was only updated to 8.57.1 because 9.0.0 and beyond deprecates the eslintrc format that's currently used by this project.

@shvlv is working on an ESlint upgrade to v9 in https://github.com/WordPress/gutenberg/pull/65648.

How did you test all changes applied to the `@wordpress/scripts` package?

Commit:
Use package-up and read-package-up forks

SmushyTaco commented:
@gziolo the tests seem to be failing from `package-up` and `read-package-up`'s dependencies only supporting ES6 modules. I've forked the base modules to support CommonJS and will work on their dependencies now to fix this. In the future I'll make sure to keep 1 PR for 1 problem like you suggested. Thankfully the work is almost done with this PR.

Regarding how I've tested all of this, I'll go into depth on this once I make the finishing touches. It's currently 1:30 AM for me so I'll aim to wrap this up after I wake up.

Commit:
CommonJS fixes and better imports

SmushyTaco commented:
I'll resolve the merge conflict once I'm home.

Commit:
Match Upstream

SmushyTaco commented:
I'll make some finishing touches to wrap this up soon

Commit:
Read Package Up Fix

Commit:
CommonJS to ESM

Commit:
ESLint Fix

SmushyTaco commented:
It looks like we need the ESLint 9 update to fix this so the unit tests pass.

Commit:
ESLint Fix

Commit:
Merge branch 'trunk' into patch-1

SmushyTaco commented:
@gziolo from what I'm seeing, it seems like the main issue is Jest not picking up the tests now that they're ESM. Jest's ESM support is still experimental and from what I see [here](https://github.com/WordPress/gutenberg/blob/trunk/test/unit/scripts/resolver.js#L20), it seems like ESM migration is something you guys want in the future. Have you guys considered migrating away from Jest and to Node's Test Runner?

With all that said, I'm going to open a new PR just for the RTLCSS bug fix soon so we can get that merged ASAP.

This PR will focus on ESM migration and keeping dependencies up to date. It'll be able to be merged after the ESLint 9 migration and after tests are migrated to ESM.

SmushyTaco commented:
See #68493

shvlv commented:
Regarding ESLint v9, the last state is we support `^8.57.0 || ^9.11.0` in `@wordress/scripts`. So, ensuring we can still support (within Gutenberg and as a standalone tool) v8 here would be nice.

Regarding the PR scope, it's hard for me to understand why we need to switch to ESM modules and why only the `scripts` package; what about the rest of the packages? The same regarding multiple dependency updates at once, I agree with @gziolo:

> In practice, it's usually easier to land targeted PRs that solve one problem at a time

Commit:
Merge Conflict Fix

SmushyTaco commented:
@shvlv yeah given the added complexity I see why solving one problem at a time can simplify things. That's why I opened #68493 which focuses just on the RTLCSS fix like this one should've. I migrated to ESM on this branch because many of the packages used here became ESM only in later versions and ESM module support from CommonJS is still experimental. It might be better to just wait for that stabilize before updating dependencies rather than migrating to ESM right now which is a much larger undertaking. With that said, I'll be closing this PR in favor of #68493

If you could look at the new PR to see why the unit tests are failing that would be greatly appreciated. The changes are minimal and I'm not seeing anything related to the changes in the error logs, which makes me think these tests were failing before any of my changes. When I run `npm run test:unit` on `trunk` with the latest stable NodeJS version (v22.13.0) and the latest version of NPM (11.0.0) the unit tests have some fails too.

So with all that said, I think #68493 should be able to be merged now to fix support for RTLCSS.
